File: java/src/main/java/it/giannibombelli/wsc2026/booking/application/integration/payment/adapter/PaymentResult.java
package it.giannibombelli.wsc2026.booking.application.integration.payment.adapter;

import it.giannibombelli.wsc2026.booking.application.commands.BookingConfirmationCommands;
import it.giannibombelli.wsc2026.booking.domain.booking.Booking;
import it.giannibombelli.wsc2026.booking.domain.booking.BookingId;
import it.giannibombelli.wsc2026.booking.domain.ports.BookingRepository;
import it.giannibombelli.wsc2026.common.utils.Require;
import it.giannibombelli.wsc2026.payment.integration.PaymentResultIntegrationEvent;

import java.util.UUID;

import static it.giannibombelli.wsc2026.booking.application.commands.BookingConfirmationCommands.ConfirmBooking;
import static it.giannibombelli.wsc2026.booking.application.commands.BookingConfirmationCommands.RejectBooking;

/**
 * Anti-Corruption Layer che traduce la Published Language esposta da {@code payment}
 * nei command interni del BC {@code booking}.
 * <p>
 * Isola il modello di {@code booking} da eventuali evoluzioni del modello interno di {@code payment}:
 * l'unico contratto condiviso è {@link PaymentResultIntegrationEvent}.
 */
public final class PaymentResult {

    private final BookingRepository bookingRepository;

    public PaymentResult(BookingRepository bookingRepository) {
        Require.requireDependency(bookingRepository, "bookingRepository");
        this.bookingRepository = bookingRepository;
    }

    public BookingConfirmationCommands adapt(PaymentResultIntegrationEvent event) {
        Require.requireArgument(event, "Payment result integration event");
        return switch (event) {
            case PaymentResultIntegrationEvent.PaymentAcceptedIntegrationEvent it -> buildConfirmBooking(UUID.fromString(it.clientReference()), it);
            case PaymentResultIntegrationEvent.PaymentRejectedIntegrationEvent it -> buildRejectBooking(UUID.fromString(it.clientReference()), it);
            case PaymentResultIntegrationEvent.PaymentExpiredIntegrationEvent ignored -> null;
        };
    }

    private BookingConfirmationCommands buildConfirmBooking(UUID bookingId, PaymentResultIntegrationEvent.PaymentAcceptedIntegrationEvent event) {
        Booking booking = findBooking(bookingId);
        if (booking == null) {
            return null;
        }
        return new ConfirmBooking(new BookingId(bookingId), booking.giftCardReference(), event.amount());
    }

    private BookingConfirmationCommands buildRejectBooking(UUID bookingId, PaymentResultIntegrationEvent.PaymentRejectedIntegrationEvent event) {
        Booking booking = findBooking(bookingId);
        if (booking == null) {
            return null;
        }
        return new RejectBooking(new BookingId(bookingId), booking.giftCardReference(), event.amount());
    }

    private Booking findBooking(UUID bookingId) {
        return bookingRepository.findById(new BookingId(bookingId)).orElse(null);
    }
}

File: typescript/src/booking/application/integration/payment/adapter/paymentResult.ts
// Anti-Corruption Layer che traduce la Published Language esposta da `payment`
// nei command interni del BC `booking`.
//
// Isola il modello di `booking` da eventuali evoluzioni del modello interno di `payment`:
// l'unico contratto condiviso è `PaymentResultIntegrationEvent`.

import type { PaymentResultIntegrationEvent } from '@/payment/integration/paymentResultIntegrationEvent.js';
import type { BookingConfirmationCommand } from '@/booking/application/commands/bookingConfirmationCommands.js';
import { confirmBooking, rejectBooking } from '@/booking/application/commands/bookingConfirmationCommands.js';
import { BookingId } from '@/booking/domain/booking/bookingId.js';
import type { BookingRepository } from '@/booking/domain/ports/bookingRepository.js';
import { Uuid } from '@/common/domain/primitive/uuid.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class PaymentResult {
  private readonly _bookingRepository: BookingRepository;

  constructor(bookingRepository: BookingRepository) {
    requireDependency(bookingRepository, 'bookingRepository');
    this._bookingRepository = bookingRepository;
  }

  adapt(event: PaymentResultIntegrationEvent): BookingConfirmationCommand | null {
    requireArgument(event, 'Payment result integration event');

    switch (event.kind) {
      case 'PaymentAcceptedIntegrationEvent':
        return this.buildConfirmBooking(event);
      case 'PaymentRejectedIntegrationEvent':
        return this.buildRejectBooking(event);
      case 'PaymentExpiredIntegrationEvent':
        return null;
    }
  }

  private buildConfirmBooking(event: PaymentResultIntegrationEvent & { kind: 'PaymentAcceptedIntegrationEvent' }): BookingConfirmationCommand | null {
    const bookingId = Uuid.fromString(event.clientReference);
    const booking = this._bookingRepository.findById(new BookingId(bookingId));
    if (booking === null) {
      return null;
    }
    return confirmBooking(new BookingId(bookingId), booking.giftCardReference(), event.amount);
  }

  private buildRejectBooking(event: PaymentResultIntegrationEvent & { kind: 'PaymentRejectedIntegrationEvent' }): BookingConfirmationCommand | null {
    const bookingId = Uuid.fromString(event.clientReference);
    const booking = this._bookingRepository.findById(new BookingId(bookingId));
    if (booking === null) {
      return null;
    }
    return rejectBooking(new BookingId(bookingId), booking.giftCardReference(), event.amount);
  }
}
